util/whitelist/debugger_whitelist.go

```diff
@@ -0,0 +1,156 @@
+	urlPath := getDebuggerURL.String()
+	resp, err := http.Get(urlPath)
+	if err != nil {
+		logrus.WithError(err).Errorf("Fail to get url (%v)", urlPath)
+		return false, err
+	}
+	defer resp.Body.Close()
+	body, _ := ioutil.ReadAll(resp.Body)
+	var data map[string]interface{}
+	err = json.Unmarshal(body, &data)
+	if err != nil {
+		return false, err
+	}
+
+	debuggerList, ok := data["debugger"]
+	isValid := (ok == true) && (debuggerList != nil)
+	whiteListCache.Set(cacheKey, isValid, cache.DefaultExpiration)
+	return isValid, nil
+}
+
+// GetInvalidIPErrorMsg return `Method not supported` (code=-32004) if invalid. Reference:
+// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1474.md
+func GetInvalidIPErrorMsg(reqByte []byte) ([]byte, error) {
+	var reqs []ReqBody // for batch request
+	var req ReqBody    // for single request
+	if isBatchRequest(reqByte, &reqs) {
+		var resps []InvalidIPRespBody
+		for _, req := range reqs {
+			var resp InvalidIPRespBody
+			resp.ID = req.ID
+			resp.JsonRPC = req.JsonRPC
+			resp.Method = req.Method
+			resp.Error.Code = -32004
+			resp.Error.Message = "Operation not permitted"
+			resps = append(resps, resp)
+		}
+		return json.Marshal(resps)
+	} else if isSingleRequest(reqByte, &req) {
+		var resp InvalidIPRespBody
+		resp.ID = req.ID
+		resp.JsonRPC = req.JsonRPC
+		resp.Method = req.Method
+		resp.Error.Code = -32004
+		resp.Error.Message = "Operation not permitted"
+		return json.Marshal(resp)
+	}
+	logrus.Debug("invalid request reqByte: ", string(reqByte))
+	return []byte("invalid request " + string(reqByte)), nil
+}
+
+type ReqBody struct {
+	ID      interface{} `json:"id,omitempty"`
+	JsonRPC string      `json:"jsonrpc,omitempty"`
+	Method  string      `json:"method,omitempty"`
+}
+
+type InvalidIPRespBody struct {
+	ID      interface{}     `json:"id,omitempty"`
+	JsonRPC string          `json:"jsonrpc,omitempty"`
+	Method  string          `json:"method,omitempty"`
+	Error   InvalidIPErrMsg `json:"error,omitempty"`
+}
+
+type InvalidIPErrMsg struct {
+	Code    int64  `json:"code,omitempty"`
+	Message string `json:"message,omitempty"`
+}
+
+func isBatchRequest(reqByte []byte, reqs *[]ReqBody) bool {
+	err := json.Unmarshal(reqByte, reqs)
+	return err == nil
+}
+
+func isSingleRequest(reqByte []byte, req *ReqBody) bool {
+	err := json.Unmarshal(reqByte, req)
+	return err == nil
+}
+
+func GetClientIPFromRequest(r *http.Request) string {
+	logrus.Debug("proxyCount: ", proxyCount)
+	if proxyCount > 0 {
+		xForwardedFor := r.Header.Get("X-Forwarded-For")
+		xRealIP := r.Header.Get("X-Real-Ip")
+
+		if xForwardedFor != "" {
+			xForwardedForParts := strings.Split(xForwardedFor, ",")
+			// Avoid reading the user's forged request header by configuring the count of reverse proxies
+			partIndex := len(xForwardedForParts) - proxyCount
+			if partIndex < 0 {
+				partIndex = 0
+			}
+			return strings.TrimSpace(xForwardedForParts[partIndex])
+		}
+
+		if xRealIP != "" {
```

Comment:
The idea with `X-Real-Ip` is that if we set `proxyCount > 0` then the last proxy will always correctly set this field so we can trust its value?

Reply:
In the usual Nginx proxy header configuration, the `X-Real-IP` header always is equal to the $remote_addr Nginx internal variable,  not append the incoming existing header like `X-Forwarded-For`. Since each proxy will reset the `X-Real-IP` header, this value is out of the malicious' control.
```
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
proxy_set_header X-Real-IP       $remote_addr;
```

In summary, in the current implementation as long as we have X-Forwarded-For or X-Real-IP header set in the reverse proxy correctly, the attacker cannot forge IP on the client side.

BTW if the proxy count is greater than 1, the X-Real-IP header value in the golang app is actually the last proxy' client IP, not the end user's IP.  This means that if it is a multi-level proxy X-Real-IP is not as effective as X-Forwarded-For.

Reply:
One configuration risk note, the IP can be forged if we are not using a reverse proxy but the proxy count is set to >1

Reply:
why not use `X-Real-IP` directly instead of checking `X-Forwarded-For` first?

Reply:
> 

As mentioned above, `X-Real-IP` does not work well in multi-level proxies env.

Reply:
i see.

Reply:
`X-Real-IP` removed. After a discussion with @iczc, `X-Forwarded-For` must be set by nginx when `proxyCount > 0`